docs/rfcs/0001-integration-with-dune.md

```diff
@@ -0,0 +1,169 @@
+- add support to produce JavaScript files in two steps: first generate `.cmo`
+files from source files, then generate `.js` files from `.cmo` files (see
+"`js.` targets" section below)
+
+Changes required in Dune:
+
+- add a new mode `melange` to `executable` stanza, e.g.
+```dune
+(executable (name foo) (modes melange))
+```
+- when building under `melange` mode, generate `cma` targets as directories with
+artifacts inside (same way as it's done with `cmxa`)
+- when building under `melange` mode, add `-bs-stop-after-cmj` to
+`cmo`-generation rules
+- when building under `melange` mode, add rules to generate `js` files from
+`cmo` files (see "`js.` targets" section below)
+- add an option, maybe as part of new stanza `melange`, to define which
+JavaScript module format will be used (`es6` or `commonjs`), e.g.:
+```dune
+(executable
+    (name foo)
+    (modes melange)
+    (melange (module es6))
+)
+```
+- add an option to choose which suffix will be added to JavaScript files
+(`.bs.js` is the default), e.g.:
+```dune
+(executable
+    (name foo)
+    (modes melange)
+    (melange (suffix bs.js))
+)
+```
+- add implicit handling targets with extension `.bs.js`, which will generate the
+necessary build rules behind the scenes, same as it is done currently for
+`bc.js` files in Js_of_ocaml.
+
+### `.js` targets
+
+One of the main challenges of Melange integration is that originally `bsc`
+command would generate both `cmo` and `js` targets with same command, but they
+need to be placed in separate directories. This does not play well with Dune
+invariants, that require all targets of a single rule
+[to be placed in same directory](https://github.com/ocaml/dune/blob/0490fd413b81cca473ef5e4fb0a6c876e000e953/src/dune_engine/rule.ml#L96-L110).
+
+After [some recent changes](https://github.com/melange-re/melange/pull/199),
+[@EduardoRFS](https://github.com/EduardoRFS) unlocked the possibility to
+generate `js` files from `cmo` files, which allows to define `cmo` and `js`
+artifacts with separate Dune rules. This allows to generate just `cmo` files
+first:
+
+```bash
+ocamlc -o output.cmo -bs-stop-after-cmo -impl output.ml
+```
+
+And then generate `js` files:
+
+```bash
+ocamlc -o output.js output.cmo
+```
+
+One remaining question is how to resolve the `import` / `require` statements in
+resulting JavaScript files, as they require knowledge about the relative paths
+to each one of the dependent modules that a module depends on.
+
+Right now, Melange [checks if the `js` file exist in the relative path](https://github.com/jchavarri/melange/blob/6524091f126ec720f7fb59873d18bdff9b001be2/jscomp/core/js_name_of_module_id.ml#L175) to get absolute paths to dependent modules, so in
+order to solve the issue we could include both `cmo` and `js` paths when calling
+`ocamlc`:
+```bash
+ocamlc -I where_your_cm_stuff_is -I where_your_js_is exe.cmo
+```
+
+But there might better approaches, like passing all load paths to dependencies
+from the command line argument.
```

Comment:
This would be kind of ugly but it would also lead to a overall simpler build pipeline for the compiler. We should still support the `-I` tho.